FeatureExtraction/Texture/texture_features.py

```diff
@@ -0,0 +1,412 @@
+                np.divide(single_glcm, denominator))
+    return results
+
+
+def calculate_IDN(glcm: GLCM):
+    if not isinstance(glcm, GLCM):
+        raise TypeError("glcm should be an instance of GLCM class")
+    y_shape, x_shape = glcm.shape[0:2]
+    x_indexes = np.linspace(0, x_shape - 1, x_shape)
+    y_indexes = np.linspace(0, y_shape - 1, y_shape)
+    cols, rows = np.meshgrid(x_indexes, y_indexes)
+    distances = glcm.shape[2]
+    angles = glcm.shape[3]
+    results = np.zeros((distances, angles))
+    for distance in range(0, distances):
+        for angle in range(0, angles):
+            single_glcm = glcm[:, :, distance, angle].reshape(y_shape, x_shape)
+            denominator = np.absolute(np.subtract(rows, cols))
+            denominator = np.divide(denominator, glcm.levels)
+            denominator = np.add(denominator, 1)
+            results[distance, angle] = np.sum(
+                np.divide(single_glcm, denominator))
+    return results
+
+
+def calculate_max_proba(glcm: GLCM):
+    if not isinstance(glcm, GLCM):
+        raise TypeError("glcm should be an instance of GLCM class")
+    distances = glcm.shape[2]
+    angles = glcm.shape[3]
+    results = np.zeros((distances, angles))
+    for distance in range(0, distances):
+        for angle in range(0, angles):
+            results[distance, angle] = np.amax(glcm[:, :, distance, angle])
+    return results
+
+
+def calculate_sum_average(glcm: GLCM):
+    if not isinstance(glcm, GLCM):
+        raise TypeError("glcm should be an instance of GLCM class")
+    y_shape, x_shape = glcm.shape[0:2]
+    distances = glcm.shape[2]
+    angles = glcm.shape[3]
+    results = np.zeros((distances, angles))
+    x_indexes = np.linspace(1, x_shape, x_shape)
+    y_indexes = np.linspace(1, y_shape, y_shape)
+    cols, rows = np.meshgrid(x_indexes, y_indexes)
+    for distance in range(0, distances):
+        for angle in range(0, angles):
+            result = 0
+            single_glcm = glcm[:, :, distance, angle].reshape(
+                (y_shape, x_shape))
+            for i in range(2, 2 * glcm.levels + 1):
+                result += i * np.sum(single_glcm[cols + rows == i])
+            results[distance, angle] = copy(result)
+    return results
+
+
+def calculate_sum_entropy(glcm: GLCM):
+    if not isinstance(glcm, GLCM):
+        raise TypeError("glcm should be an instance of GLCM class")
+    y_shape, x_shape = glcm.shape[0:2]
+    distances = glcm.shape[2]
+    angles = glcm.shape[3]
+    results = np.zeros((distances, angles))
+    x_indexes = np.linspace(1, x_shape, x_shape)
+    y_indexes = np.linspace(1, y_shape, y_shape)
+    cols, rows = np.meshgrid(x_indexes, y_indexes)
+    for distance in range(0, distances):
+        for angle in range(0, angles):
+            result = 0
+            single_glcm = glcm[:, :, distance, angle].reshape(
+                (y_shape, x_shape))
+            for i in range(2, 2 * glcm.levels + 1):
```

Comment:
Why do you start from 2? Maybe you can call this number somehow so this loop would be easier to understand.

Reply:
I start from 2 because the formula says so. Included a constant.